diff -uNr samba-3.0.7.orig/source/Makefile.in samba-3.0.7/source/Makefile.in
--- samba-3.0.7.orig/source/Makefile.in	2004-10-25 10:06:08.000000000 -0700
+++ samba-3.0.7/source/Makefile.in	2004-10-25 10:03:31.000000000 -0700
@@ -1371,6 +1371,7 @@
 
 python_clean:
 	@-if test -n "$(PYTHON)"; then $(PYTHON) python/setup.py clean; fi
+	@-rm -rf build/
 
 # revert to the previously installed version
 revert:
@@ -1437,6 +1438,7 @@
 	@rm -f include/proto.h include/build_env.h  include/wrepld_proto.h \
 		nsswitch/winbindd_proto.h web/swat_proto.h \
 		client/client_proto.h utils/net_proto.h \
+	        utils/ntlm_auth_proto.h \
 		smbd/build_options.c
 
 MKPROTO_SH = $(srcdir)/script/mkproto.sh
@@ -1512,6 +1514,7 @@
 
 distclean: realclean
 	-rm -f include/stamp-h
+	-rm -f script/gen-8bit-gap.sh smbadduser
 	-rm -f include/config.h Makefile
 	-rm -f config.status config.cache so_locations
 	-rm -rf .deps TAGS
