--- a/Makefile.in	2008-11-18 01:38:34.000000000 +0300
+++ b/Makefile.in	2009-09-05 09:50:00.000000000 +0400
@@ -256,6 +256,9 @@ ifeq ($(UCLIBC_FORMAT_SHARED_FLAT),y)
 		fi; \
 	done
 endif
+# Install precursors for later optimization
+	$(INSTALL) -d $(PREFIX)$(DEVEL_PREFIX)lib/optinfo
+	$(INSTALL) -m 644 lib/libc.a lib/interp.os $(PREFIX)$(DEVEL_PREFIX)lib/optinfo/
 
 # Installs run-time libraries
 install_runtime: all
