--- ppp.orig/pppd/pathnames.h	2008-11-26 13:02:29.000000000 +0300
+++ ppp/pppd/pathnames.h	2009-02-09 18:08:09.000000000 +0300
@@ -9,40 +9,40 @@
 
 #else /* HAVE_PATHS_H */
 #ifndef _PATH_VARRUN
-#define _PATH_VARRUN 	"/etc/ppp/"
+#define _PATH_VARRUN 	"/var/run/"
 #endif
 #define _PATH_DEVNULL	"/dev/null"
 #endif /* HAVE_PATHS_H */
 
 #ifndef _ROOT_PATH
-#define _ROOT_PATH
+#define _ROOT_PATH	"/tmp"
 #endif
 
-#define _PATH_UPAPFILE 	 _ROOT_PATH "/etc/ppp/pap-secrets"
-#define _PATH_CHAPFILE 	 _ROOT_PATH "/etc/ppp/chap-secrets"
-#define _PATH_SRPFILE 	 _ROOT_PATH "/etc/ppp/srp-secrets"
-#define _PATH_SYSOPTIONS _ROOT_PATH "/etc/ppp/options"
-#define _PATH_IPUP	 _ROOT_PATH "/etc/ppp/ip-up"
-#define _PATH_IPDOWN	 _ROOT_PATH "/etc/ppp/ip-down"
-#define _PATH_IPPREUP	 _ROOT_PATH "/etc/ppp/ip-pre-up"
-#define _PATH_AUTHUP	 _ROOT_PATH "/etc/ppp/auth-up"
-#define _PATH_AUTHDOWN	 _ROOT_PATH "/etc/ppp/auth-down"
-#define _PATH_TTYOPT	 _ROOT_PATH "/etc/ppp/options."
-#define _PATH_CONNERRS	 _ROOT_PATH "/etc/ppp/connect-errors"
-#define _PATH_PEERFILES	 _ROOT_PATH "/etc/ppp/peers/"
-#define _PATH_RESOLV	 _ROOT_PATH "/etc/ppp/resolv.conf"
+#define _PATH_UPAPFILE 	 _ROOT_PATH "/ppp/pap-secrets"
+#define _PATH_CHAPFILE 	 _ROOT_PATH "/ppp/chap-secrets"
+#define _PATH_SRPFILE 	 _ROOT_PATH "/ppp/srp-secrets"
+#define _PATH_SYSOPTIONS _ROOT_PATH "/ppp/options"
+#define _PATH_IPUP	 _ROOT_PATH "/ppp/ip-up"
+#define _PATH_IPDOWN	 _ROOT_PATH "/ppp/ip-down"
+#define _PATH_IPPREUP	 _ROOT_PATH "/ppp/ip-pre-up"
+#define _PATH_AUTHUP	 _ROOT_PATH "/ppp/auth-up"
+#define _PATH_AUTHDOWN	 _ROOT_PATH "/ppp/auth-down"
+#define _PATH_TTYOPT	 _ROOT_PATH "/ppp/options."
+#define _PATH_CONNERRS	 _ROOT_PATH "/var/log/ppp-connect-errors"
+#define _PATH_PEERFILES	 _ROOT_PATH "/ppp/peers/"
+#define _PATH_RESOLV	 _ROOT_PATH "/ppp/resolv.conf"
 
 #define _PATH_USEROPT	 ".ppprc"
 #define	_PATH_PSEUDONYM	 ".ppp_pseudonym"
 
 #ifdef INET6
-#define _PATH_IPV6UP     _ROOT_PATH "/etc/ppp/ipv6-up"
-#define _PATH_IPV6DOWN   _ROOT_PATH "/etc/ppp/ipv6-down"
+#define _PATH_IPV6UP     _ROOT_PATH "/ppp/ipv6-up"
+#define _PATH_IPV6DOWN   _ROOT_PATH "/ppp/ipv6-down"
 #endif
 
 #ifdef IPX_CHANGE
-#define _PATH_IPXUP	 _ROOT_PATH "/etc/ppp/ipx-up"
-#define _PATH_IPXDOWN	 _ROOT_PATH "/etc/ppp/ipx-down"
+#define _PATH_IPXUP	 _ROOT_PATH "/ppp/ipx-up"
+#define _PATH_IPXDOWN	 _ROOT_PATH "/ppp/ipx-down"
 #endif /* IPX_CHANGE */
 
 #ifdef __STDC__
@@ -51,13 +51,13 @@
 #ifdef HAVE_PATHS_H
 #define _PATH_PPPDB	"/var/run/pppd2.tdb"
 #else
-#define _PATH_PPPDB	"/etc/ppp/pppd2.tdb"
+#define _PATH_PPPDB	"/var/run/pppd2.tdb"
 #endif
 #endif /* __STDC__ */
 
 #ifdef PLUGIN
 #ifdef __STDC__
-#define _PATH_PLUGIN	DESTDIR "/lib/pppd/" VERSION
+#define _PATH_PLUGIN	"/usr/lib/pppd/"
 #else /* __STDC__ */
 #define _PATH_PLUGIN	"/usr/lib/pppd"
 #endif /* __STDC__ */
