--- pppd-2.4.5/pppd/ipv6cp.c	2005-11-04 10:20:14.000000000 +0000
+++ ppp/pppd/ipv6cp.c	2012-02-09 13:59:25.751543002 +0000
@@ -179,6 +179,8 @@ int no_ifaceid_neg = 0;
 
 /* local vars */
 static int ipv6cp_is_up;
+static char *path_ipv6up = _PATH_IPV6UP;	/* pathname of ipv6-up script */
+static char *path_ipv6down = _PATH_IPV6DOWN;	/* pathname of ipv6-down script */
 
 /*
  * Callbacks for fsm code.  (CI = Configuration Information)
@@ -255,6 +257,11 @@ static option_t ipv6cp_option_list[] = {
     { "ipv6cp-max-failure", o_int, &ipv6cp_fsm[0].maxnakloops,
       "Set max #conf-naks for IPv6CP", OPT_PRIO },
 
+    { "ipv6-up-script", o_string, &path_ipv6up,
+      "Set pathname of ipv6-up script", OPT_PRIV },
+    { "ipv6-down-script", o_string, &path_ipv6down,
+      "Set pathname of ipv6-down script",OPT_PRIV },
+
    { NULL }
 };
 
@@ -1294,7 +1301,7 @@ ipv6cp_up(f)
      */
     if (ipv6cp_script_state == s_down && ipv6cp_script_pid == 0) {
 	ipv6cp_script_state = s_up;
-	ipv6cp_script(_PATH_IPV6UP);
+	ipv6cp_script(path_ipv6up);
     }
 }
 
@@ -1345,7 +1352,7 @@ ipv6cp_down(f)
     /* Execute the ipv6-down script */
     if (ipv6cp_script_state == s_up && ipv6cp_script_pid == 0) {
 	ipv6cp_script_state = s_down;
-	ipv6cp_script(_PATH_IPV6DOWN);
+	ipv6cp_script(path_ipv6down);
     }
 }
 
@@ -1388,13 +1395,13 @@ ipv6cp_script_done(arg)
     case s_up:
 	if (ipv6cp_fsm[0].state != OPENED) {
 	    ipv6cp_script_state = s_down;
-	    ipv6cp_script(_PATH_IPV6DOWN);
+	    ipv6cp_script(path_ipv6down);
 	}
 	break;
     case s_down:
 	if (ipv6cp_fsm[0].state == OPENED) {
 	    ipv6cp_script_state = s_up;
-	    ipv6cp_script(_PATH_IPV6UP);
+	    ipv6cp_script(path_ipv6up);
 	}
 	break;
     }
