diff -ur ppp-2.4.2/pppd/auth.c pppd/pppd/auth.c
--- ppp-2.4.2/pppd/auth.c	2003-06-12 03:56:26.000000000 +0400
+++ pppd/pppd/auth.c	2005-01-12 22:18:24.637707136 +0300
@@ -1050,6 +1051,10 @@
     if (idle_time_hook != 0) {
 	tlim = idle_time_hook(&idle);
     } else {
+/* JYWeng 20031216: replace itime with idle.xmit_idle for only outgoing traffic is counted*/
+	if(tx_only) 
+		itime = idle.xmit_idle;
+	else
 	itime = MIN(idle.xmit_idle, idle.recv_idle);
 	tlim = idle_time_limit - itime;
     }
diff -ur ppp-2.4.2/pppd/options.c pppd/pppd/options.c
--- ppp-2.4.2/pppd/options.c	2004-01-13 07:02:07.000000000 +0300
+++ pppd/pppd/options.c	2005-01-12 22:18:24.690699080 +0300
@@ -65,6 +65,7 @@
 #if defined(ultrix) || defined(NeXT)
 char *strdup __P((char *));
 #endif
+bool tx_only;			/* JYWeng 20031216: idle time counting on tx traffic */
 
 static const char rcsid[] = RCSID;
 
@@ -309,6 +310,10 @@
       "Check for traffic limit every N seconds", OPT_PRIO | OPT_LLIMIT | 1 },
 #endif
 
+/* JYWeng 20031216: add for tx_only option*/
+    { "tx_only", o_bool, &tx_only,
+      "set idle time counting on tx_only or not", 1 },
+
     { NULL }
 };
 
diff -ur ppp-2.4.2/pppd/pppd.h pppd/pppd/pppd.h
--- ppp-2.4.2/pppd/pppd.h	2003-04-07 04:01:46.000000000 +0400
+++ pppd/pppd/pppd.h	2005-01-12 22:18:24.732692696 +0300
@@ -209,6 +209,7 @@
 /*
  * Global variables.
  */
+extern bool tx_only;			/* JYWeng 20031216: idle time counting on tx traffic */
 
 extern int	hungup;		/* Physical layer has disconnected */
 extern int	ifunit;		/* Interface unit number */
